File: 待办清单功能/ruankai2/app/src/main/java/com/byted/camp/todolist/ui/NoteListAdapter.java
package com.byted.camp.todolist.ui;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Build;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewParent;

import com.byted.camp.todolist.NoteOperator;
import com.byted.camp.todolist.R;
import com.byted.camp.todolist.TodoListApplication;
import com.byted.camp.todolist.beans.Note;
import com.byted.camp.todolist.beans.State;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Comparator;
import java.util.List;


public class NoteListAdapter extends RecyclerView.Adapter<NoteViewHolder> {

    private NoteOperator operator;
    private final List<Note> notes = new ArrayList<Note>();

    public NoteListAdapter(NoteOperator operator) {
        this.operator = operator;
    }





    public void refresh(List<Note> newNotes) {
            notes.clear();
            if (newNotes != null) {
                notes.addAll(newNotes);
//            notes.sort(Comparator.comparing());
                SharedPreferences sp = TodoListApplication.getContext().getSharedPreferences("sort",Context.MODE_MULTI_PROCESS);
                boolean isOpen = sp.getBoolean("isopen",false);

                Log.d("adapter", "isOpen:"+isOpen);
                if(isOpen){

                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                    //升序排序
                    notes.sort(new Comparator<Note>() {
                        @Override
                        //返回正值即前值大于后值
                        public int compare(Note n1, Note n2) {
                            if (n1.getState() != n2.getState()) {
                                //完成state==1，未完成state为0，即完成的放在未完成的后面
                                return State.refrom(n1.getState()) - State.refrom(n2.getState());
                            } else {
                                if (n1.getPriority() != n2.getPriority()) {
                                    //优先级由1到3递减
                                    return n2.getPriority() - n1.getPriority();
                                } else {
                                    return n2.getDate().compareTo(n1.getDate());
                                }
                            }
                        }
                    });
                } else
                    Log.e("data", "Build.VERSION.SDK_INT < Build.VERSION_CODES.N");
                for (Note item : notes) {
                    Log.d("data", "In NoteListAdapter id:" + item.getId() + " state:" + item.getState() + " priority:" + item.getPriority());
                }
                //todo 根据${com.byted.camp.todolist.operation.activity.SettingActivity} 中设置的sp控制是否将已完成的完成排到最后，默认不排序
            }
        }
            notifyDataSetChanged();



    }

    @NonNull
    @Override
    public NoteViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int pos) {
        View itemView = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_note, parent, false);
        return new NoteViewHolder(itemView, operator);
    }

    @Override
    public void onBindViewHolder(@NonNull NoteViewHolder holder, int pos) {
        holder.bind(notes.get(pos));
    }

    @Override
    public int getItemCount() {
        return notes.size();
    }
}

File: 12.25整合代码/MyApplication2.2/app/src/main/java/com/example/myapplication/todolist/ui/NoteListAdapter.kt
package com.example.myapplication.todolist.ui

import android.content.Context
import android.content.SharedPreferences
import android.os.Build
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.example.myapplication.R
import com.example.myapplication.todolist.NoteOperator
import com.example.myapplication.TodoListApplication
import com.example.myapplication.todolist.beans.Note
import com.example.myapplication.todolist.beans.State
import java.util.ArrayList
import java.util.Comparator


class NoteListAdapter(private val operator: NoteOperator) :
    RecyclerView.Adapter<NoteViewHolder>() {
    private val notes: MutableList<Note> = ArrayList()
    fun refresh(newNotes: List<Note>?) {
        notes.clear()
        if (newNotes != null) {
            notes.addAll(newNotes)
            //            notes.sort(Comparator.comparing());
            val sp: SharedPreferences = TodoListApplication.context!!
                .getSharedPreferences("sort", Context.MODE_MULTI_PROCESS)
            val isOpen = sp.getBoolean("isopen", false)
            Log.d("adapter", "isOpen:$isOpen")
            if (isOpen) {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                    //升序排序
                    notes.sortWith(object : Comparator<Note?> {
                        //返回正值即前值大于后值
                        override fun compare(n1: Note?, n2: Note?): Int {
                            return if (n1!!.state !== n2!!.state&&n1!!.state!=null&&n2!!.state!=null) {
                                //完成state==1，未完成state为0，即完成的放在未完成的后面
                                State.refrom(n1!!.state!!) - State.refrom(n2!!.state!!)
                            } else {
                                if (n1!!.priority !== n2!!.priority) {
                                    //优先级由1到3递减
                                    n2!!.priority - n1!!.priority
                                } else {
                                    n2!!.date!!.compareTo(n1!!.date)
                                }
                            }
                        }


                    })
                } else Log.e("data", "Build.VERSION.SDK_INT < Build.VERSION_CODES.N")
                for (item in notes) {
                    Log.d("data",
                        "In NoteListAdapter id:" + item.id
                            .toString() + " state:" + item.state
                            .toString() + " priority:" + item.priority
                    )
                }
                //todo 根据${com.byted.camp.todolist.operation.activity.SettingActivity中设置的sp控制是否将已完成的完成排到最后
            }
        }
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, pos: Int): NoteViewHolder {
        val itemView: View = LayoutInflater.from(parent.context)
            .inflate(R.layout.item_note_todo, parent, false)
        return NoteViewHolder(itemView, operator)
    }

    override fun onBindViewHolder(holder: NoteViewHolder, pos: Int) {
        holder.bind(notes[pos])
    }

    override fun getItemCount(): Int {
        return notes.size
    }
}
